src/appsync.ts

```diff
@@ -152,419 +164,447 @@ export class AppsyncVTL extends VTL {
-   *
-   * ```ts
-   * import * as appsync from "@aws-cdk/aws-appsync-alpha";
-   *
-   * const api = new appsync.GraphQLApi(..);
-   *
-   * ```ts
-   * const getPerson = new AppsyncResolver<{id: string}, Person | undefined>(
-   *   ($context, id) => {
-   *     const person = table.get({
-   *       key: {
-   *         id: $util.toDynamoDB(id)
-   *       }
-   *     });
-   *     return person;
-   *   });
-   * ```
-   *
-   * // code first person type
-   * const personType = api.addType(appsync.ObjectType(...));
-   *
-   * api.addQuery("getPerson", getPerson.getField(api, personType))
-   * getPerson.createResolver(api, {
-   *   typeName: "Query",
-   *   fieldName: "getPerson"
-   * });
-   * ```
-   *
-   * @param api app sync API which the data sources construct will be added under.
-   * @param returnType the code-first graphql return type of this field
-   * @param options optional fields like arguments and directive to add to the schema
+   * The API this {@link AppsyncField} should be added to.
    */
-  public getField(
-    api: appsync.GraphqlApi,
-    returnType: appsync.GraphqlType,
-    options?: Omit<
-      appsync.ResolvableFieldOptions,
-      | "returnType"
-      | keyof ReturnType<
-          AppsyncResolver<Arguments, Result, Source>["getResolvableFields"]
-        >
-    >
-  ): appsync.ResolvableField {
-    const fields = this.getResolvableFields(api);
-
-    return new appsync.ResolvableField({
+  readonly api: appsync.GraphqlApi;
+}
+
+/**
+ * Generate a resolvable field to use with AppSync CDK's Code field strategy.
+ *
+ * ```ts
+ * import * as appsync from "@aws-cdk/aws-appsync-alpha";
+ *
+ * const api = new appsync.GraphQLApi(..);
+ *
+ * ```ts
+ * const getPersonField = new AppsyncField<{id: string}, Person | undefined>(
+ *   {
+ *     api,
+ *     returnType: appsync.Field.string(),
+ *     args: {
+ *       argName: appsync.Field.string()
+ *     }
+ *   },
+ *   ($context, id) => {
+ *     const person = table.get({
+ *       key: {
+ *         id: $util.toDynamoDB(id)
+ *       }
+ *     });
+ *     return person;
+ *   });
+ * ```
+ *
+ * // code first person type
+ * const personType = api.addType(appsync.ObjectType(...));
+ *
+ * api.addQuery("getPerson", getPersonField)
+ * ```
+ *
+ * @param api app sync API which the data sources construct will be added under.
+ * @param returnType the code-first graphql return type of this field
+ * @param options optional fields like arguments and directive to add to the schema
+ */
+export class AppsyncField<
+  Arguments extends ResolverArguments,
+  Result,
+  Source = undefined
+> extends appsync.ResolvableField {
+  /**
+   * This static property identifies this class as an AppsyncResolver to the TypeScript plugin.
+   */
+  public static readonly FunctionlessType = "AppsyncField";
+
+  constructor(
+    options: AppsyncFieldOptions,
+    resolve: ResolverFunction<Arguments, Result, Source>
+  ) {
```

Comment:
I was trying to remember why field and resolver are different.

Ha...

https://github.com/aws/aws-cdk/issues/18351#issuecomment-1115266100

I don't like how buggy it is, ran into this again when working on the authorizer example.

Can you add a validator or something that checks to see if a resolver was actually created?

Reply:
Do you mean a node validator? There is no `Construct` here for me to add that validator to. Also, if the resolver isn't added, then the validator also wouldn't be added? Is this solved with just tests?

Reply:
Hmm, you could put it on the api.

```ts
api.node.addValidator(() => {
    api.node.tryFindChild(`${props.typeName}${props.fieldName}Resolver`) ? [] : ["uhh ohh no resolver!"]
});
```

Reply:
I don't have a `typeName` or `fieldName` at this point. That happens later when you call `addQuery`. I don't think there's anything to validate here?